packages/store/addon/-private/identifiers/cache.ts

```diff
@@ -0,0 +1,426 @@
+    this._generate = configuredGenerationMethod || defaultGenerationMethod;
+    this._update = configuredUpdateMethod || defaultEmptyCallback;
+    this._forget = configuredForgetMethod || defaultEmptyCallback;
+    this._reset = configuredResetMethod || defaultEmptyCallback;
+  }
+
+  // allows us to peek without generating when needed
+  // useful for the "create" case when we need to see if
+  // we are accidentally overwritting something
+  peekRecordIdentifier(resource: ResourceIdentifierObject, shouldGenerate: true): StableRecordIdentifier;
+  peekRecordIdentifier(resource: ResourceIdentifierObject, shouldGenerate: false): StableRecordIdentifier | undefined;
+  peekRecordIdentifier(
+    resource: ResourceIdentifierObject,
+    shouldGenerate: boolean = false
+  ): StableRecordIdentifier | undefined {
+    // short circuit if we're already the stable version
+    if (isStableIdentifier(resource)) {
+      if (DEBUG) {
+        // TODO should we instead just treat this case as a new generation skipping the short circuit?
+        if (!(resource.lid in this._cache.lids) || this._cache.lids[resource.lid] !== resource) {
+          throw new Error(`The supplied identifier ${resource} does not belong to this store instance`);
+        }
+      }
+      return resource;
+    }
+
+    // `type` must always be present
+    if (DEBUG) {
+      if (!isNonEmptyString(resource.type)) {
+        throw new Error('resource.type needs to be a string');
+      }
+    }
+
+    let type = normalizeModelName(resource.type);
+    let keyOptions = getTypeIndex(this._cache.types, type);
+    let identifier: StableRecordIdentifier | undefined;
+    let lid = coerceId(resource.lid);
+    let id = coerceId(resource.id);
+
+    // go straight for the stable RecordIdentifier key'd to `lid`
+    if (lid !== null) {
+      identifier = keyOptions.lid[lid];
+    }
+
+    // we may have not seen this resource before
+    // but just in case we check our own secondary lookup (`id`)
+    if (identifier === undefined && id !== null) {
+      identifier = keyOptions.id[id];
+    }
+
+    if (identifier === undefined) {
+      // we have definitely not seen this resource before
+      // so we allow the user configured `GenerationMethod` to tell us
+      let newLid = this._generate(resource, 'record');
+
+      // we do this _even_ when `lid` is present because secondary lookups
+      // may need to be populated, but we enforce not giving us something
+      // different than expected
+      if (lid !== null && newLid !== lid) {
+        throw new Error(`You should not change the <lid> of a RecordIdentifier`);
+      } else if (lid === null) {
+        // allow configuration to tell us that we have
+        // seen this `lid` before. E.g. a secondary lookup
+        // connects this resource to a previously seen
+        // resource.
+        identifier = keyOptions.lid[newLid];
+      }
+
+      if (shouldGenerate === true) {
```

Comment:
maybe `if (shouldGenerate)` ?

Reply:
I generally prefer stricter checks (no type casts) with TS both will narrow but one advantage of the stricter runtime check is that folks can't get away with passing in "truthy" and "falsey" things in practice. We've historically had issues with "accepting anything" and then trying to either case or weed out all the bad inputs in EmberData, with a few instances where it turned out we were converting things from objects to null (such as during createRecord) that caused us issues refactoring later.